python/lsst/faro/utils/stats_utils.py

```diff
@@ -0,0 +1,85 @@
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import numpy as np
+from lsst.pipe.base import Struct
+from scipy.stats import median_abs_deviation
+
+__all__ = (
+    "calcQuartileClippedStats",
+)
+
+
+def calcQuartileClippedStats(dataArray, nSigmaToClip=3.0):
+    """Calculate the quartile-based clipped statistics of a data array.
+    The difference between quartiles[2] and quartiles[0] is the interquartile
+    distance.  0.74*interquartileDistance is an estimate of standard deviation
+    so, in the case that ``dataArray`` has an approximately Gaussian
+    distribution, this is equivalent to nSigma clipping.
+    Parameters
+    ----------
+    dataArray : `list` or `numpy.ndarray` of `float`
+        List or array containing the values for which the quartile-based
+        clipped statistics are to be calculated.
+    nSigmaToClip : `float`, optional
+        Number of \"sigma\" outside of which to clip data when computing the
+        statistics.
+    Returns
+    -------
+    result : `lsst.pipe.base.Struct`
+        The quartile-based clipped statistics with ``nSigmaToClip`` clipping.
+        Atributes are:
+        ``median``
+            The median of the full ``dataArray`` (`float`).
+        ``mean``
+            The quartile-based clipped mean (`float`).
+        ``stdDev``
+            The quartile-based clipped standard deviation (`float`).
+        ``rms``
+            The quartile-based clipped root-mean-squared (`float`).
+        ``clipValue``
+            The value outside of which to clip the data before computing the
+            statistics (`float`).
+        ``goodArray``
+            A boolean array indicating which data points in ``dataArray`` were
+            used in the calculation of the statistics, where `False` indicates
+            a clipped datapoint (`numpy.ndarray` of `bool`).
+    """
+    quartiles = np.percentile(dataArray, [25, 50, 75])
+    assert len(quartiles) == 3
+    median = quartiles[1]
+    interQuartileDistance = quartiles[2] - quartiles[0]
+    clipValue = nSigmaToClip * 0.74 * interQuartileDistance
+    good = np.logical_not(np.abs(dataArray - median) > clipValue)
+    quartileClippedMean = dataArray[good].mean()
+    quartileClippedStdDev = dataArray[good].std()
+    quartileClippedRms = np.sqrt(np.mean(dataArray[good] ** 2))
+    quartileClippedMad = median_abs_deviation(dataArray[good])
```

Comment:
This is the MAD not the Sigma MAD? For the plots we report the Sigma MAD so that it is more comparable to the STD.

Reply:
I think what you call "sigma MAD" is what I have always known as "MAD". For example, in `analysis.drp.colorColorFitPlot.py` you use `from scipy.stats import median_absolute_deviation as sigmaMad`. I simply didn't rename it in `faro`.

Reply:
Hmm, and I expect `sigmaMAD` = 1.4826 `MAD`

Reply:
That's new to me. What is "sigmaMAD" supposed to be? (i.e., a definition -- I've never heard the phrase "sigmaMAD" except from Princeton pipelines folks... ;) )

Reply:
Sigma MAD is the MAD multiplied by a constant that makes it comparable to the sigma derived from a gaussian. You can read about it here: https://en.wikipedia.org/wiki/Median_absolute_deviation

Reply:
This seems to imply that there's _another_ thing we should settle -- do we want to report rms (as in Ivezic+2004), MAD, sigmaMAD, or something else? (Note that analysis_drp seems to be reporting MAD, but calling it sigmaMAD.) We basically need to decide what we want to adopt and stick with it.

Reply:
The median_absolute_deviation function from scipy includes the scaling factor as default, hence why I imported it as sigmaMAD which I felt was less confusing than the default name. The median_abs_deviation from scipy does not include the scaling factor as default so to get the sigma you need to either set scale="normal" or scale=1/1.4826.

Reply:
Note that median_absolute_deviation is deprecated, so we should use median_abs_deviation (with "scale='normal'") instead. (see https://docs.scipy.org/doc/scipy/reference/stats.html)